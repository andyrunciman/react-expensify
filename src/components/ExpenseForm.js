import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component{
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100 ).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt):moment(),
            calendarFocused:false,
            error:''  
        }
    }
    onNoteChange = (event) => {
        const note = event.target.value;
        this.setState(()=>({note}));
    }
    onDescriptionChange = (event) =>{
        const description = event.target.value;
        this.setState(()=>({description}));
    }
    onAmountChange = (event) =>{
        const amount = event.target.value;        
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}));
        }
    }
    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(()=>({createdAt}));
        } 
    }
    onFocusChange = ({focused}) => {
        this.setState(()=>({calendarFocused:focused}));
    }
    onSubmit = (event) => {
        event.preventDefault();
        if(!this.state.description || !this.state.amount){
            const error = "Please complete the description and amount";
            this.setState(()=>({error}))
        }else{
            this.setState(()=>({error:''}))
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10)*100, //as in pense
                createdAt:this.state.createdAt.valueOf(),
                note:this.state.note
            });
        }
    }
    render(){
        return (
            <div>
                <form className="expense-form__form" onSubmit={this.onSubmit}>
                    {this.state.error?<p>{this.state.error}</p>:undefined}
                    <div className="input-item">
                        <input 
                            type="text" 
                            className="input-text"
                            placeholder="Description"
                            autoFocus
                            value={this.state.description}
                            onChange={this.onDescriptionChange}/>
                    </div>
                    <div className="input-item">
                        <input
                            type="text"
                            placeholder="Amount"
                            className="input-text"
                            value={this.state.amount}
                            onChange={this.onAmountChange}/>
                    </div>
                    <SingleDatePicker
                        date = {this.state.createdAt}
                        onDateChange ={this.onDateChange}
                        focused = {this.state.calendarFocused}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths = {1}
                        isOutsideRange = {()=> false}
                    />
                    <div className="input-item">
                        <textarea
                            placeholder="Add a note for your expense"
                            className="input-text"
                            value={this.state.note}
                            onChange={this.onNoteChange}>
                        </textarea>
                    </div>
                    <button className="btn btn--primary">Add Expense</button>
                </form>
            </div>
        )
    }
}