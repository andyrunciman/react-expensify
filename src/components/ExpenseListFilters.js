// Instead of the mapDispatchToProps function we can use a shorthand.
// First, we import all the action creators as an object from actions/filters:
// import * as filterActions from "../actions/filters"; 
// and in the connect function we pass that object as the second argument:
// export default connect(mapStateToProps, filterActions)(ExpenseListFilter); 
// And that's it, we now have on our props all the action creators. 
// More on this: Dan Abramov on egghead.io

import React from 'react';
import {setTextFilter,sortByDate,sortByAmount,setStartDate,setEndDate} from '../actions/filters';
import {connect} from 'react-redux';
import 'react-dates/initialize';
import {DateRangePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export class ExpenseListFilters extends React.Component{
    state = {
        calendarFocused:null
    };
    onDatesChange = ({startDate,endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onFocusChange = (c) => {
        this.setState(()=>({calendarFocused:c}));
    }
    render(){
        return (
            <div className="expense-list-filter">
                <div className ="expense-list-filter__box">
                  
                        <div className = 'expense-list-filter__item'>
                            <input className="input-text" type='text' placeholder = "Search expenses" value={this.props.filters.text} onChange={(event)=>{
                                this.props.setTextFilter(event.target.value);
                            }}/>
                        </div>  
                  
                  
                        <div className = 'expense-list-filter__item'>
                            <select  className="input-combo" value={this.props.filters.sortBy} onChange={(event)=>{
                                if(event.target.value==="date"){
                                    this.props.sortByDate();
                                }else{
                                    this.props.sortByAmount();
                                }
                            }}>
                                <option value="date">Date</option>
                                <option value="amount">Amount</option>
                            </select>
                            </div>  
                    
                  
                        <div className = 'expense-list-filter__item'>
                            <DateRangePicker
                                startDate = {this.props.filters.startDate}
                                startDateId= {"start1"}
                                endDate = {this.props.filters.endDate}
                                endDateId =  {"end1"}
                                onDatesChange = {this.onDatesChange}
                                focusedInput = {this.state.calendarFocused}
                                onFocusChange = {this.onFocusChange}
                                numberOfMonths = {1}
                                showClearDates = {true} //shows the clear button
                                isOutsideRange = {()=>false}    
                            />
                        </div>  
                    
            </div>
        </div>
        )
    }
}

const matchDispatchToProps = (dispatch) => {
    return {
        setStartDate:(startDate) => dispatch(setStartDate(startDate)),
        setEndDate:(endDate) => dispatch(setEndDate(endDate)),
        setTextFilter:(filter) => dispatch(setTextFilter(filter)),
        sortByDate:() => dispatch(sortByDate()),
        sortByAmount:() => dispatch(sortByAmount()),
    }
}
const mapStateToProps = (state)=>{
    return{
        filters:state.filters
    }
}
export default connect(mapStateToProps,matchDispatchToProps )(ExpenseListFilters);