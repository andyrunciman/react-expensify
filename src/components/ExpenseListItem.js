import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({id,description,amount,createdAt,dispatch})=>(
    <div className='expense-list-item'>
        
        <div>
            <Link className="expense-list-item__title" to={`/edit/${id}`}>{description}</Link>
            <h3 className="expense-list-item__date">{moment(createdAt).format('Do, MMM YYYY')}</h3>
        </div>

        <div >
            <h3 className="expense-list-item__amount">{numeral(amount/100).format('$0,0.00')}</h3>
        </div>
        
            
            
        
    </div>
)

export default ExpenseListItem;