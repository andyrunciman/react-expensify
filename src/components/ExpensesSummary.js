import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({numberOfExpenses,expensesTotal}) => {
    const expenseText = numberOfExpenses>1?'expenses':'expense';
    const formattedAmount = numeral(expensesTotal/100).format('$0,0.00');
    return (
            <div className="expense-summary">
                <div className="expense-summary__box">
                    <p className="expense-summary__text u-margin-bottom-small ">Viewing <span>{numberOfExpenses}</span> {expenseText} totalling <span>{formattedAmount}</span></p>
                    <Link className="btn btn--primary-link" to="/create">Add Expense</Link>
                </div>
            </div>
            
        );
}

const mapStateAsProps = (state) => {
    const filteredExpenses = getVisibleExpenses(state.expenses,state.filters);
    return {
        expensesTotal:getExpensesTotal(filteredExpenses),
        numberOfExpenses:filteredExpenses.length
    }
}
export default connect(mapStateAsProps)(ExpensesSummary);
