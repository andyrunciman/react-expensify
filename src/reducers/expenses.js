const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState ,action) => {
    switch(action.type){
        case 'ADD_EXPENSE':
            return [...state,action.expense];
        case 'DELETE_EXPENSE':
            return action.id?state.filter(item=>item.id!==action.id):state;
        case 'EDIT_EXPENSE':
            //const expense = {...state.find(id=>id===action.id),...action.expense}
            console.log('editing expense inside reducer',action.updates)
            return state.map((expense)=>{
                if(expense.id === action.id){
                    return {
                        ...expense,
                        ...action.updates
                    }
                }else{
                    return expense;
                }
            });
        default:
            return state;
    }
}

export default expensesReducer;