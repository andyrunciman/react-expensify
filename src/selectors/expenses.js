import Moment from 'moment';
const getVisibleExpenses = (expenses,{text,sortBy,startDate,endDate}) => {
    return expenses.filter((expense)=>{
        const createdAtMoment = Moment(expense.createdAt);
        const filterText = expense.description.toLowerCase().includes(text.toLowerCase());
        //const filterStart = typeof startDate !== 'number' || expense.createdAt >=startDate;
        //const filterEnd   = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const filterStart = startDate ? startDate.isSameOrBefore(createdAtMoment,'day'): true
        const filterEnd = endDate ? endDate.isSameOrAfter(createdAtMoment,'day'): true
        return filterText && filterStart && filterEnd;
    }).sort((expense1,expense2)=>{
        switch(sortBy){
            case 'description':
                return expense1.description > expense2.description?-1:1;
            case 'date':
                return expense1.createdAt > expense2.createdAt?-1:1;
            case 'amount':
                return expense1.amount > expense2.amount?-1:1;
            default:
                return 1;
        }
    });
}

export default getVisibleExpenses;