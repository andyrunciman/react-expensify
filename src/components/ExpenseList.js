import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';


export const ExpenseList = (props) => (
    <div className="expense-list">
        <div className="expense-list__box">
            <div className ="expense-list__header">
                <h3>Expense</h3>
                <h3>Amount</h3>
            </div>
            {props.expenses.length === 0
                ?<div className="expense-list-item">No expenses</div>
                :props.expenses.map(expense=><ExpenseListItem key={expense.id} {...expense}/>)
            }
        </div>
        
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses:selectExpenses(state.expenses,state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList)
