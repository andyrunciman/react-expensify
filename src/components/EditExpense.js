import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startRemoveExpense,startEditExpense} from '../actions/expenses';

export class EditExpense extends React.Component{
    
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id,expense);
        this.props.history.push('/');
    }
    onDelete = () => {
        this.props.removeExpense(this.props.expense.id);
        this.props.history.push('/');
    }
    render(){
        return(
            <div className='expense-form'>
                <div className='expense-form__box'>
                    <ExpenseForm onSubmit={this.onSubmit} expense={this.props.expense}/>
                    <button className="btn btn--primary" onClick={this.onDelete}>Remove</button>
                </div>
            </div>
            
        )
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        editExpense: (id,expense) => dispatch(startEditExpense(id,expense)),
        removeExpense: (id) => dispatch(startRemoveExpense(id))
    }
}

const mapStateToProps = (state,props) => {
    return {
        expense:state.expenses.find((expense)=>expense.id === props.match.params.id)
    };
}

export default connect(mapStateToProps,mapDispatchToProps)(EditExpense);