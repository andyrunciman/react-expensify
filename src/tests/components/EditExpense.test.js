import React from 'react';
import {shallow} from 'enzyme';
import {EditExpense} from '../../components/EditExpense';
import expenses from '../../tests/fixtures/expenses';

let wrapper,history,removeExpense,editExpense;

beforeEach(()=>{
    history = {push:jest.fn()};
    editExpense = jest.fn();
    removeExpense = jest.fn();
    wrapper = shallow(<EditExpense  removeExpense={removeExpense} 
                                    editExpense={editExpense} 
                                    history={history} 
                                    expense={expenses[1]}/>);
});
test('should render EditExpensePage',()=>{
    expect(wrapper).toMatchSnapshot();
});
test('should handle edit expense',()=>{
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id,expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
});
test('should handle removeExpense',()=>{
    wrapper.find('button').simulate('click');
    //note - the tutor says that button elements dont have props so you should call the simulate. My code worked OK without them so I am not sure. It does make sense that the simulate actutally simulates a click which then calls the function, rather than just calling the click function.
    expect(removeExpense).toHaveBeenLastCalledWith(expenses[1].id);
    expect(history.push).toHaveBeenLastCalledWith('/');

});