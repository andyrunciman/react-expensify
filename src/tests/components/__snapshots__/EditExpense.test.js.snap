// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div
  className="expense-form"
>
  <div
    className="expense-form__box"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 109500,
          "createdAt": -1000,
          "description": "Rent",
          "id": "2",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="btn btn--primary"
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</div>
`;
